src/Services/CouponService.php

```diff
@@ -2,30 +2,312 @@
+        if(isset($basket['couponCode']) && strlen($basket['couponCode']) > 0)
+        {
+            $campaign = $this->couponCampaignRepository->findByCouponCode($basket['couponCode']);
+
+            if($campaign instanceof CouponCampaign)
+            {
+                if($campaign->couponType == CouponCampaign::COUPON_TYPE_SALES)
+                {
+                    $basket['openAmount']       = $basket['basketAmount'];
+                    $basket["basketAmount"]     -= $basket['couponDiscount'];
+                    $basket["basketAmountNet"]  -= $basket['couponDiscount'];
+
+                }
+                $basket['couponCampaignType'] = $campaign->couponType;
+            }
+        }
+        return $basket;
+    }
+
+    /**
+     * Validate the basket for the coupon, and remove the coupon if invalid
+     * @param Basket $basket
+     * @param array $basketItem Current basketItem
+     */
+    public function validateBasketItemDelete($basket, $basketItem) {
+        if(strlen($basket->couponCode) > 0)
+        {
+            $campaign = $this->couponCampaignRepository->findByCouponCode($basket->couponCode);
+
+            if($campaign instanceof CouponCampaign)
+            {
+                if($this->isCouponMinimalOrderOnDelete($basket, $basketItem, $campaign))
+                {
+                    // Check if the minimal order value is not met
+                    $this->removeInvalidCoupon(CouponService::BELOW_MINIMAL_ORDER_VALUE);
+                }
+                else if($this->isCouponValidForBasketItems($basket, $basketItem, $campaign))
+                {
+                    // Check if the coupon is still valid with the new basket (only coupon item removed?)
+                    $this->removeInvalidCoupon(CouponService::NO_VALID_ITEM_IN_BASKET);
+                }
+            }
+        }
+    }
+
+    /**
+     * Validate the basket for the coupon, and remove the coupon if invalid
+     * @param Basket $basket
+     * @param array $data New basketItem
+     * @param array $basketItem Current basketItem
+     */
+    public function validateBasketItemUpdate($basket, $data, $basketItem) {
+        if(strlen($basket->couponCode) > 0)
+        {
+            $campaign = $this->couponCampaignRepository->findByCouponCode($basket->couponCode);
+
+            if($campaign instanceof CouponCampaign)
+            {
+                if($this->isCouponMinimalOrderOnUpdate($basket, $data, $basketItem, $campaign))
+                {
+                    // Check if the minimal order value is not met
+                    $this->removeInvalidCoupon(CouponService::BELOW_MINIMAL_ORDER_VALUE);
+                }
+            }
+        }
+    }
+
+    /**
+     * Remove a coupon and throw a notification
+     * @param int $code Infocode for notification
+     */
+    private function removeInvalidCoupon($code)
+    {
+        $this->removeCoupon();
+        pluginApp(NotificationService::class)->info('CouponValidation', $code);
```

Comment:
Kann die eine Zeile nicht in die removeCoupon dazu ?

Reply:
removeCoupon hat eine Verwendung in der CouponResource, wo der return von removeCouponCode verwendet wird. 
Erfordert Prüfung

Reply:
done